File: Examples/Counter/ios/Counter/Strohm/Subscriptions.swift
import Foundation

class Subscriptions {
    let strohm: Strohm
    var pendingSubscriptions: [() -> Void]? = []
    var subscribers: [UUID: HandlerFunction] = [:]

    init(strohm: Strohm) {
        self.strohm = strohm
        strohm.comms.registerHandlerFunction(name: "subscriptionUpdate",
                                             function: self.subscriptionUpdateHandler)
    }

    func addSubscriber(propsSpec: PropsSpec,
                       handler: @escaping HandlerFunction,
                       completion: @escaping (UUID) -> Void) {
        if pendingSubscriptions != nil {
            pendingSubscriptions!.append({ [weak self] in
                self?.subscribe_(
                    propsSpec: propsSpec,
                    handler: handler,
                    completion: completion)
            })
        } else {
            subscribe_(propsSpec: propsSpec,
                       handler: handler,
                       completion: completion)
        }
    }

    private func subscribe_(propsSpec: PropsSpec,
                            handler: @escaping HandlerFunction,
                            completion: (UUID) -> Void) {
        guard let encodedPropsSpec = strohm.comms.encode(object: propsSpec) else {
            return
        }
        let subscriptionId = UUID()
        subscribers[subscriptionId] = handler
        strohm.call(method: "strohm.store.subscribe_from_native(\"\(subscriptionId.uuidString)\", \"\(encodedPropsSpec)\")")
        completion(subscriptionId)
    }

    func removeSubscriber(subscriptionId: UUID) {
        subscribers.removeValue(forKey: subscriptionId)
        strohm.call(method: "strohm.store.unsubscribe_from_native(\"\(subscriptionId.uuidString)\")")
    }

    func effectuatePendingSubscriptions() {
        if let pending = pendingSubscriptions {
            pendingSubscriptions = nil
            for subscriber in pending { subscriber() }
        }
    }

    func handlePropsUpdate(props: Props, subscriptionId: UUID) {
        subscribers[subscriptionId]?(props)
    }

    func subscriptionUpdateHandler(args: JsonComms.Arguments) {
        if let subscriptionIdString = args["subscriptionId"] as? String,
           let subscriptionId = UUID(uuidString: subscriptionIdString),
           let newProps = args["new"] as? [String: Any] {
            handlePropsUpdate(props: newProps,
                              subscriptionId: subscriptionId)
        }
    }
}

File: Strohm/android/strohm/src/main/java/dev/strohmnative/strohm/Subscriptions.kt
package dev.strohmnative.strohm

import kotlinx.collections.immutable.PersistentList
import kotlinx.collections.immutable.persistentListOf
import java.util.UUID
import java.util.concurrent.atomic.AtomicMarkableReference

class Subscriptions {
    val strohm: Strohm

    internal var _pendingSubscriptions: AtomicMarkableReference<PersistentList<() -> Unit>>
    private var subscribers: MutableMap<UUID, HandlerFunction>

    internal constructor(strohm: Strohm) {
        this.strohm = strohm
        this._pendingSubscriptions = AtomicMarkableReference(persistentListOf(), true)
        this.subscribers = mutableMapOf()
        strohm.comms.registerHandlerFunction("subscriptionUpdate") { args ->
            this.subscriptionUpdateHandler(args)
        }
    }

    fun addSubscriber(
        propsSpec: PropsSpec,
        handler: HandlerFunction,
        completion: (UUID) -> Unit
    ) {
        var succeeded = false
        while (!succeeded) {
            val holder = BooleanArray(1)
            val pending = _pendingSubscriptions.get(holder)
            val isUsingPending = holder[0]

            if (isUsingPending) {
                val newPending = pending.add { this.subscribe_(propsSpec, handler, completion) }
                succeeded = _pendingSubscriptions.compareAndSet(pending, newPending, true, true)
            } else {
                subscribe_(propsSpec, handler, completion)
                succeeded = true
            }
        }
    }

    private fun subscribe_(
        propsSpec: PropsSpec,
        handler: HandlerFunction,
        completion: (UUID) -> Unit
    ) {
        val subscriptionId = UUID.randomUUID()
        subscribers[subscriptionId] = handler
        val encodedPropsSpec = strohm.comms.encode(propsSpec)
        strohm.call("strohm.native.subscribe_from_native(\"$subscriptionId\", \"$encodedPropsSpec\")")
        completion(subscriptionId)
    }

    internal fun removeSubscriber(subscriptionId: UUID) {
        subscribers.remove(subscriptionId)
        strohm.call("strohm.native.unsubscribe_from_native(\"$subscriptionId\")")
    }

    internal fun effectuatePendingSubscriptions() {
        var isUsingPending: Boolean
        var pending: PersistentList<() -> Unit>
        var clearSucceeded: Boolean
        do {
            val holder = BooleanArray(1)
            pending = _pendingSubscriptions.get(holder)
            isUsingPending = holder[0]

            clearSucceeded = _pendingSubscriptions.compareAndSet(pending, persistentListOf(), true, false)
        } while (!clearSucceeded && isUsingPending)

        if (isUsingPending) {
            pending.forEach { it() }
        }
    }

    internal fun handlePropsUpdate(props: Props, subscriptionId: UUID) {
        subscribers[subscriptionId]?.let { it(props) }
    }

    internal fun subscriptionUpdateHandler(args: CommsHandlerArguments) {
        val subscriptionIdString = args["subscriptionId"] as? String
        @Suppress("UNCHECKED_CAST")
        val newProps = args["new"] as? Props
        if (subscriptionIdString != null && newProps != null) {
            val subscriptionId = UUID.fromString(subscriptionIdString)
            handlePropsUpdate(newProps, subscriptionId)
        }
    }
}

